README
#*********************************************************#
#                                                         #
#     README for check_rhev3 plugin for Icinga/Nagios     #
#                                                         #
#*********************************************************#

What is check_rhev3?
====================

This plugin for Icinga/Nagios is used to monitor a variety of
an RHEV or oVirt environement including datacenters, clusters, hosts,
vms, vm pools and storage domains.


Contributions / Help
====================

Send email to rkoch@rk-it.at if you have questions regarding use
of this software. Please include the following information in your email:
* OS type and version
* Perl version

To submit patches or suggest improvements, send email to rkoch@rk-it.at or create
a pull request on Github. Thanks.


Installation
============

The check_rhev3 installation can be used by following the standard autotools
installation process, documented in the INSTALL file. As a quick start you can do

$ ./configure
$ make all
# make install

This will install check_rhev3 to /usr/local/nagios/libexec and the PNP template
to /usr/local/pnp4nagios/share/templates.

To change these paths use the following options:
--prefix=<path_to_nagios_plugins>
--with-pnp-dir=<path_to_pnp_share_folder>

If you don't use PNP you can disable installing the PNP template with:
--disable-pnp-template

To change the user and group ownership of these files use:
--with-nagios-user=<user>
--with-nagios-group=<group>

The 'nagios-plugins-rhev3.spec' file demonstrates how to distribute check_rhev3
as an RPM package.

Also make sure that the following requirements are met:
Perl
Perl-Modules:
  LWP::UserAgent
  perl-Crypt-SSLeay
  HTTP::Request
  Getopt::Long
  XML:Simple

For Fedora 19 (and newer) install the following packages:
# yum install perl-libwww-perl perl-LWP-Protocol-https perl-HTTP-Message \
              perl-XML-Simple

For RHEL 6 install the following packages:
# yum install perl-Crypt-SSLeay perl-libwww-perl perl-XML-Simple

For Debian Squeeze install the following packages:
# apt-get install libwww-perl libcrypt-ssleay-perl libxml-simple-perl


Documentation
=============

Short documentation on how to use this plugin.

Detailed documentation can be found on the project website
https://github.com/ovido/check_rhev3.

The following checks can be performed by this plugin:

Datacenter
----------
check_rhev3 -H rhevm -a admin@internal:password -D default
check_rhev3 -H rhevm -a admin@internal:password -D default -l status
check_rhev3 -H rhevm -a admin@internal:password -D default -l version
check_rhev3 -H rhevm -a admin@internal:password -D default -l storage
check_rhev3 -H rhevm -a admin@internal:password -D default -l storage -s status
check_rhev3 -H rhevm -a admin@internal:password -D default -l storage -s usage
check_rhev3 -H rhevm -a admin@internal:password -D default -l storage -s overall-usage

Cluster
-------
check_rhev3 -H rhevm -a admin@internal:password -C default
check_rhev3 -H rhevm -a admin@internal:password -C default -l hosts
check_rhev3 -H rhevm -a admin@internal:password -C default -l vms
check_rhev3 -H rhevm -a admin@internal:password -C default -l networks

Hosts
-----
check_rhev3 -H rhevm -a admin@internal:password -R rhevh
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l status
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l load
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l cpu
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l cpu -s usage
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l ksm
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l memory
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l memory -s mem
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l memory -s swap
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l network
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l network -s status
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l network -s traffic
check_rhev3 -H rhevm -a admin@internal:password -R rhevh -l network -s errors

Storage
-------
check_rhev3 -H rhevm -a admin@internal:password -S storage
check_rhev3 -H rhevm -a admin@internal:password -S storage -l status
check_rhev3 -H rhevm -a admin@internal:password -S storage -l usage

VMs
---
check_rhev3 -H rhevm -a admin@internal:password -M vm
check_rhev3 -H rhevm -a admin@internal:password -M vm -l status
check_rhev3 -H rhevm -a admin@internal:password -M vm -l cpu
check_rhev3 -H rhevm -a admin@internal:password -M vm -l memory
check_rhev3 -H rhevm -a admin@internal:password -M vm -l network
check_rhev3 -H rhevm -a admin@internal:password -M vm -l network -s traffic
check_rhev3 -H rhevm -a admin@internal:password -M vm -l network -s errors

VM Pools
--------
check_rhev3 -H rhevm -a admin@internal:password -P pool
check_rhev3 -H rhevm -a admin@internal:password -P pool -l usage


Sample Icinga/Nagios definitions
--------------------------------

Detailed information can be found on the project website
https://github.com/ovido/check_rhev3

Command
-------
define command{
  command_name check_rhev3
  command_line $USER1$/check_rhev3 -H $_RHEVM$ -a $ARG1$ $ARG2$
}

Host
----
define host{
  use       linux-server
  host_name rhevh
  alias     RHEV Hypervisor
  address   192.168.1.2
  _rhevm    192.168.1.1
}

Service
-------
define service{
  use                 generic-service
  host_name           rhevh
  service_description RHEV CPU Check
  check_command       check_rhev3!admin@internal:password!-R $HOSTNAME$ -l cpu
}



ChangeLog
#*********************************************************#
#                                                         #
#    Changelog of check_rhev3 plugin for Icinga/Nagios    #
#                                                         #
#*********************************************************#

* 1.6 - Sun Nov 06 2016
- New features:
-   Add support for oVirt 4 and RHEV 4 (#42)
-   Add option to ignore xml warnings (#49)
-   Components in state maintenance result in warning state now instead of critical
- Bugs fixed:
-   Critical/Warning thresholds for -l vms/hosts are ignored (#44)

* 1.5 - Fri Feb 19 2016
- New features:
-   Set cluster to warning if host is in maintenance mode instead of critical (Issue #27)
-   New storage domain check: -s overall-usage (Issue #28)
-   Display more detailed information of host and vm status (Issue #35)
Bugs fixed:
-   Wrong CPU Usage on hypervisor maintenance (Issue #26)
-   Discrepancy in DataCenter Storage usage Output. (Issue #28) 
-   Wrong CPU value for vms (Issue #29)
-   VM CPU usage graphs show values over 100% (Issue #33)
-   Unsuccessful stat on filename containing newline (Issue #36)
-   align behaviour for check_statistics cpu to memory (Issue #39)

* 1.4 - Wed Apr 16 2014
- New features:
-   Get status of vms on a specific host (Issue #8)
-   Proxy/No Proxy environment variable support (Issue #12)
-   Removed RHEV password from debug output and added plugin version number (Issue #13)
-   Default port changed to 443 (Issue #15)
-   Improved source code formatting (Issue #16)
-   Default warning and critical thresholds for CPU load based on number of cores (Issue #17)
-   More detailed status information for datacenters, hosts, vms and storagedomains (Issue #18)
-   Display network traffic in Mbit/s in PNP4Nagios (Issue #20)
-   PNP4Nagios templates for datacenter, hosts and vms stati (Issue #23)
- Bugs fixed:
-   Interface traffic in Mbytes/s instead of Bytes/s (Issue #14)
-   Authentication fails if password contains equal sign (Issue #21)
-   Initialization of variable in check_rhev3.pl - thanks Administratoor (Issue #22)
-   Memory usage of vms in hypervisor set to 0 if negative (Issue #24)

* 1.3 - Tue Nov 07 2013
- New features:
-   Cookie based authentication as default auth method (Issue #1)
-   Use fixed values for warning and critical values of storage domans (Issue #3)
-   Network interface error monitoring (Issue #4)
-   Improved status messages for NICs, storage (Issue #7)

- Bugs fixed:
-    Improved filter regex for NICs (Issue #2)
-    Optimized error messages (Issue #6)
-    Memory usage fix (Issue #10)

* 1.2.1 - Tue Jul 23 2013
- Bug fixed:
-   Fixed search regex for host/network search

* 1.2 - Thu May 16 2013
- General:
-   Moved project to github: https://github.com/ovido/check_rhev3
-
- New features:
-   Verify RHEV-M certificate (Ticket #23)
-   Allow authentication sessions for authentication in RHEV >= 3.1 and oVirt >= 3.1 (Ticket #24)
-   Use option -n <nic> to check a specific nic (Ticket #25)
-
- Bugs fixed:
-   Performance data issue with check_multi (Ticket #9) 

* 1.1 - Wed Jan 30 2013
- New features:
-   Support of RHEV 3.1 and oVirt 3.2 (Ticket #2, #3, #4)
-   Use same regex as in RHEV Manager search field (Ticket #4)
-   Authentication via authrc file (Ticket #6)
-
- Bugs fixed:
-    Storage domain search fixed for RHEV 3.1 (Ticket #3)
-    Network traffic check fixed for RHEV 3.1 (Ticket #2)
-    Adapt BuildRequires in nagios-plugins-rhev3.spec (Ticket #5)
-    Fix LWP::UserAgent settings for Fedora 18 (Ticket #8)

* 1.0.1 - Thu Dec 27 2012
- Scan for LWP::UserAgent version before establishing HTTPS
  connection to RHEV Manager (Bug #1)

* 1.0 - Fri Aug 31 2012
- This is the first stable release of new plugin check_rhev3

* 0.1.1 - Fri Aug 17 2012
- Removed Nagios::Plugin dependency

* 0.1.0 - Mon Aug 10 2012
- This is the first public beta release of new plugin check_rhev3



contrib/check_rhev3.php
<?php
#
# Plugin: check_rhev3
# Author: Rene Koch <rkoch@rk-it.at>
# Created: 2012/08/14
# Last update: 2015/02/07
#

if ($NAME[1] == "cpu"){

  # process CPU statistics
  $opt[1] = "--vertical-label \"CPU utilization\" -l 0 --title \"CPU utilization for $hostname\" --slope-mode -u 100 -N";
  $def[1]  = "DEF:var1=$RRDFILE[1]:$DS[1]:AVERAGE ";
  $def[1] .= "DEF:var2=$RRDFILE[1]:$DS[2]:AVERAGE ";
  $def[1] .= "DEF:var3=$RRDFILE[1]:$DS[3]:AVERAGE ";

  if ($NAME[2] == "cpu.current.user"){

    # process hypervisor CPU stats
    $def[1] .= "DEF:var4=$RRDFILE[1]:$DS[4]:AVERAGE ";
    $def[1] .= "CDEF:sp2=var2 ";
    $def[1] .= "CDEF:sp3=var3 ";
    $def[1] .= "CDEF:sp4=var4 ";

    $def[1] .= "AREA:sp2#000080:\"User    \" ";
    $def[1] .= "GPRINT:sp2:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp2:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp2:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";
    $def[1] .= "STACK:sp3#FFCC00:\"System  \" ";
    $def[1] .= "GPRINT:sp3:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp3:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp3:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";
    $def[1] .= "STACK:sp4#CCCCCC:\"Idle    \" ";
    $def[1] .= "GPRINT:sp4:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp4:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp4:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";

  }else{

    # process VM CPU stats
    $def[1] .= "CDEF:sp1=100,var1,- ";
    $def[1] .= "CDEF:sp2=var1 ";
    $def[1] .= "CDEF:sp3=var3 ";

    $def[1] .= "AREA:sp2#000080:\"Guest      \" ";
    $def[1] .= "GPRINT:sp2:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp2:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp2:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";
    $def[1] .= "STACK:sp3#FFCC00:\"Hypervisor \" ";
    $def[1] .= "GPRINT:sp3:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp3:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp3:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";
    $def[1] .= "STACK:sp1#CCCCCC:\"Idle       \" ";
    $def[1] .= "GPRINT:sp1:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp1:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:sp1:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";

  }

}elseif (preg_match("/traffic/i",$NAME[1])){

  # proccess network traffic
  $opt[1] = "--vertical-label \"Traffic usage\" -l 0 --title \"Traffic utilization for $hostname\" --slope-mode -N";
  $def[1] = "";

  foreach ($this->DS as $key=>$val){
    $ds = $val['DS'];
    $def[1] .= "DEF:var$key=$RRDFILE[$ds]:$ds:AVERAGE ";
    $def[1] .= "CDEF:traffic$key=var$key,8,* ";
    $def[1] .= "LINE1:traffic$key#" . color() . ":\"" . substr($LABEL[$ds],8) ."      \" ";
    $def[1] .= "GPRINT:traffic$key:LAST:\"last\: %3.4lgMbit/s \" ";
    $def[1] .= "GPRINT:traffic$key:MAX:\"max\: %3.4lgMbit/s \" ";
    $def[1] .= "GPRINT:traffic$key:AVERAGE:\"average\: %3.4lgMbit/s \"\\n ";
  }

}elseif (preg_match("/_up/i", $NAME[1])){
	
  # process service status
  $opt[1] = "--vertical-label \"" . str_replace('_up', '', $NAME[1]) . " status\" --slope-mode -N";
  $def[1] = "";
  
  foreach ($this->DS as $key=>$val){
  	$components = array("Datacenters_", "Clusters_", "Hosts_", "Vms_", "Storagedomains_");
  	$ds = $val['DS'];
  	$def[1] .= "DEF:var$key=$RRDFILE[$ds]:$ds:AVERAGE ";
  	$def[1] .= "LINE1:var$key#" . color() . ":\"" . str_replace($components, '', $LABEL[$ds]) . "     \" ";
  	$def[1] .= "GPRINT:var$key:LAST:\"last\: %3.4lg \" ";
  	$def[1] .= "GPRINT:var$key:MAX:\"max\: %3.4lg \" ";
  	$def[1] .= "GPRINT:var$key:AVERAGE:\"average\: %3.4lg \"\\n ";
  }

}elseif (preg_match("/storage_/i", $NAME[1])){

  # process storage domain usage
  $opt[1] = "--vertical-label \"" . "Storage usage\" --slope-mode -N";
  $def[1] = "";

  foreach ($this->DS as $key=>$val){
  	$ds = $val['DS'];
  	$def[1] .= "DEF:var$key=$RRDFILE[$ds]:$ds:AVERAGE ";
  	$def[1] .= "LINE1:var$key#" . color() . ":\"" . str_replace('storage_', '', $LABEL[$ds]) . "     \" ";
  	$def[1] .= "GPRINT:var$key:LAST:\"last\: %3.4lg \" ";
  	$def[1] .= "GPRINT:var$key:MAX:\"max\: %3.4lg \" ";
  	$def[1] .= "GPRINT:var$key:AVERAGE:\"average\: %3.4lg \"\\n ";
  }

}else{

  # process load, ksm, memory and swap statistics
  if ($NAME[1] == "ksm.cpu.current"){
    $opt[1] = "--vertical-label \"KSM utilization\" -l 0 --title \"KSM utilization for $hostname\" --slope-mode -u 100 -N";
  }elseif ($NAME[1] == "cpu.load.avg.5m"){
    $opt[1] = "--vertical-label \"load utilization\" -l 0 --title \"Load utilization for $hostname\" --slope-mode -N";
  }elseif ($NAME[1] == "memory"){
    $opt[1] = "--vertical-label \"Memory\" -l 0 --title \"Memory utilization for $hostname\" --slope-mode -u 100 -N";
  }elseif ($NAME[1] == "swap"){
    $opt[1] = "--vertical-label \"Swap\" -l 0 --title \"Swap utilization for $hostname\" --slope-mode -u 100 -N";
  }elseif (($NAME[1] == "Hosts") || ($NAME[1] == "hosts")){
    $opt[1] = "--vertical-label \"Hosts up\" -l 0 --title \"Hosts with status UP\" --slope-mode -N";
  }elseif ($NAME[1] == "nics"){
    $opt[1] = "--vertical-label \"NICs active\" -l 0 --title \"NICs with status Active\" --slope-mode -N";
  }elseif ($NAME[1] == "networks"){
    $opt[1] = "--vertical-label \"Networks operational\" -l 0 --title \"Networks with status Operational\" --slope-mode -N";
  }elseif ($NAME[1] == "Datacenters"){
    $opt[1] = "--vertical-label \"Datacenters up\" -l 0 --title \"Datacenters with status UP\" --slope-mode -N";
  }elseif ($NAME[1] == "storagedomains"){
    $opt[1] = "--vertical-label \"Storagedomains active\" -l 0 --title \"Storagedomains with status Active\" --slope-mode -N";
  }elseif ($NAME[1] == "vmpool"){
    $opt[1] = "--vertical-label \"VMs up\" -l 0 --title \"VMs used in Pool\" --slope-mode -N";
  }elseif ($NAME[1] == "Vms"){
    $opt[1] = "--vertical-label \"VMs up\" -l 0 --title \"VMs with status UP\" --slope-mode -N";
  }elseif (preg_match("/storage_/i",$NAME[1])){
    $opt[1] = "--vertical-label \"Disk utilization\" -l 0 --title \"Storage utilization\" --slope-mode -u 100 -N";
  }else{
    $opt[1] = "--vertical-label \"" . strtoupper($NAME[1]) . "\" -l 0 --title \"" . strtoupper($NAME[1]) . " utilization for $hostname\" --slope-mode -N";
  }

  if ( ($NAME[1] == "ksm.cpu.current") || ($NAME[1] == "memory") || ($NAME[1] == "cpu.load.avg.5m") || ($NAME[1] == "swap") ){

    $def[1]  = "DEF:var1=$RRDFILE[1]:$DS[1]:AVERAGE ";
    $def[1] .= "CDEF:sp1=var1,100,/,12,* ";
    $def[1] .= "CDEF:sp2=var1,100,/,30,* ";
    $def[1] .= "CDEF:sp3=var1,100,/,50,* ";
    $def[1] .= "CDEF:sp4=var1,100,/,70,* ";

    $def[1] .= "AREA:var1#FF5C00:\"$NAME[1]    \" ";
    $def[1] .= "AREA:sp4#FF7C00: ";
    $def[1] .= "AREA:sp3#FF9C00: ";
    $def[1] .= "AREA:sp2#FFBC00: ";
    $def[1] .= "AREA:sp1#FFDC00: ";

    $def[1] .= "GPRINT:var1:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:var1:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:var1:AVERAGE:\"average\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "LINE1:var1#000000: ";

  }else{

    $def[1]  = "DEF:var1=$RRDFILE[1]:$DS[1]:AVERAGE ";
    $def[1] .= "LINE1:var1#000000:\"" . $NAME[1] . "     \" ";
    $def[1] .= "GPRINT:var1:LAST:\"last\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:var1:MAX:\"max\: %3.4lg$UNIT[1] \" ";
    $def[1] .= "GPRINT:var1:AVERAGE:\"average\: %3.4lg$UNIT[1] \"\\n ";

  }

}


# generate html color code
function color(){
  $color = dechex(rand(0,10000000));
  while (strlen($color) < 6){
    $color = dechex(rand(0,10000000));
  }
  return $color;
}

?>
